src/FSharpVSPowerTools.Core/Utils.fs

```diff
@@ -27,94 +58,241 @@ module List =
             res.[i] <- array.[0..i]
         res
 
+
+    /// Fold over the array passing the index and element at that index to a folding function
     let foldi (folder : 'State -> int -> 'T -> 'State) (state : 'State) (array : 'T[]) =
-        let mutable state = state
+        checkNonNull "array" array
+        if array.Length = 0 then state else
+        let folder = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt folder
+        let mutable state:'State = state
         let len = array.Length
         for i = 0 to len - 1 do
-            state <- folder state i array.[i]
+            state <- folder.Invoke (state, i, array.[i])
         state
 
-    open System.Collections.Generic
+
+    /// Get the first element of the array or None if the Array is null or empty
+    let tryHead array =        
+        match array with
+        | null | [||] -> None
+        | arr  -> Some arr.[0]    
+
+
+    /// Get the last element of the array or None if the Array is null or empty
+    let tryLast array =
+        match array with
+        | null | [||] -> None
+        | arr  -> Some arr.[arr.Length-1]        
+
 
     /// Returns an array that contains no duplicate entries according to generic hash and
     /// equality comparisons on the entries.
     /// If an element occurs multiple times in the array then the later occurrences are discarded.
     let distinct (array:'T[]) =
         checkNonNull "array" array
+        if array.Length = 0 then [||] else
         let temp = Array.zeroCreate array.Length
         let mutable i = 0
-
-        let hashSet = HashSet<'T>(HashIdentity.Structural<'T>)
+        let hashSet = HashSet<'T> HashIdentity.Structural<'T>
         for v in array do 
             if hashSet.Add(v) then
                 temp.[i] <- v
                 i <- i + 1
+        temp.[0..i-1]
 
-        Array.sub temp 0 i 
 
-    /// Optimized arrays equality. ~100x faster than `array1 = array2`.
-    let inline areEqual (x: 'a[]) (y: 'a[]) =
-        match x, y with
-        | null, null -> true
-        | [||], [||] -> true
-        | null, _ | _, null -> false
-        | _ when x.Length <> y.Length -> false
-        | _ ->
-            let mutable break' = false
-            let mutable i = 0
-            let mutable result = true
-            while i < x.Length && not break' do
-                if x.[i] <> y.[i] then 
-                    break' <- true
-                    result <- false
+    let distinctBy keyf (array:'T[]) =
+        checkNonNull "array" array
+        if array.Length = 0 then [||] else
+        let temp = Array.zeroCreate array.Length
+        let mutable i = 0 
+        let hashSet = HashSet<_> HashIdentity.Structural<_>
+        for v in array do
+            if hashSet.Add(keyf v) then
+                temp.[i] <- v
                 i <- i + 1
-            result
+        temp.[0..i-1]
+
+    /// pass an array byref to reverse it in place
+    let revInPlace (array: 'a[] byref ) =
```

Comment:
You don't have to pass array by ref because arrays are ref types. Change it to `let revInPlace (array: 'a[])`.